<a href="https://colab.research.google.com/github/lalygiraudo/prueba-git/blob/main/Clasificaci%C3%B3n__binaria.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# ============================================================
# CLASIFICACIÓN BINARIA CON REGRESIÓN LOGÍSTICA
# DATASET: BREAST CANCER WISCONSIN
# ============================================================

# 1. IMPORTAR LIBRERÍAS

import pandas as pd

from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

In [ ]:
# ============================================================
# 2. CARGAR EL DATASET
# ============================================================

datos = load_breast_cancer()

X = pd.DataFrame(
    datos.data,
    columns=datos.feature_names
)

y = pd.Series(
    datos.target,
    name="diagnostico"
)

print("Cantidad de filas:", X.shape[0])
print("Cantidad de variables:", X.shape[1])

Cantidad de filas: 569
Cantidad de variables: 30


In [ ]:
# ============================================================
# 3. CONOCER LOS DATOS
# ============================================================

print("Primeras filas:")

display(X.head())

Primeras filas:


,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst radius,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,0.07871,...,25.38,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,0.05667,...,24.99,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,...,23.57,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,0.09744,...,14.91,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,0.05883,...,22.54,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678


In [ ]:
# ============================================================
# 4. CONOCER LA VARIABLE OBJETIVO
# ============================================================

print("Clases disponibles:")

print(datos.target_names)

print("\nCantidad de casos por clase:")

print(y.value_counts())

Clases disponibles:
['malignant' 'benign']

Cantidad de casos por clase:
diagnostico
1    357
0    212
Name: count, dtype: int64


0 = malignant
1 = benign

In [ ]:
# ============================================================
# 5. SEPARAR DATOS DE ENTRENAMIENTO Y PRUEBA
# ============================================================

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)

print("Datos de entrenamiento:", X_train.shape)
print("Datos de prueba:", X_test.shape)

Datos de entrenamiento: (455, 30)
Datos de prueba: (114, 30)


In [ ]:
# ============================================================
# 6. ESTANDARIZAR LAS VARIABLES
# ============================================================

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)

X_test_scaled = scaler.transform(X_test)

In [ ]:
# ============================================================
# 7. CREAR EL MODELO
# ============================================================

modelo = LogisticRegression(
    max_iter=1000,
    random_state=42
)

# Entrenar
modelo.fit(X_train_scaled, y_train)

print("Modelo entrenado correctamente.")

Modelo entrenado correctamente.


In [ ]:
# ============================================================
# 8. REALIZAR PREDICCIONES
# ============================================================

y_pred = modelo.predict(X_test_scaled)

print("Predicciones:")
print(y_pred)

print("\nValores reales:")
print(y_test.values)

Predicciones:
[0 1 0 1 0 1 1 0 0 0 1 0 1 0 0 1 0 1 1 1 0 0 1 1 1 1 0 1 1 1 1 1 1 1 0 0 1
 1 1 0 1 1 1 0 0 1 1 1 1 0 1 1 1 1 1 1 1 0 0 1 1 1 1 1 0 0 1 1 1 1 1 1 1 0
 0 0 0 1 1 1 1 1 0 1 0 1 1 1 1 1 1 1 0 0 0 1 0 1 0 1 0 0 0 1 0 0 1 0 1 0 1
 0 1 1]

Valores reales:
[0 1 0 1 0 1 1 0 0 0 1 0 1 0 0 1 1 1 1 1 0 0 1 1 1 1 0 1 1 1 1 1 1 1 0 0 1
 1 1 0 1 1 1 0 0 1 1 1 1 0 1 1 1 0 1 1 1 0 0 1 1 1 1 1 0 0 1 1 1 1 1 1 1 0
 0 0 0 1 1 1 1 1 0 1 0 1 1 1 1 1 1 1 0 0 0 1 0 1 0 1 0 0 0 1 0 0 1 0 1 0 1
 0 1 1]


In [ ]:
# ============================================================
# 9. ACCURACY
# ============================================================

accuracy = accuracy_score(y_test, y_pred)

print("Accuracy:", round(accuracy, 3))

Accuracy: 0.982


In [ ]:
# ============================================================
# 10. MATRIZ DE CONFUSIÓN
# ============================================================

matriz = confusion_matrix(y_test, y_pred)

print(matriz)

[[41  1]
 [ 1 71]]


In [ ]:
print("             Predicción")
print("             Maligno  Benigno")
print()
print("Real Maligno", matriz[0])
print("Real Benigno", matriz[1])

             Predicción
             Maligno  Benigno

Real Maligno [41  1]
Real Benigno [ 1 71]


In [ ]:
# ============================================================
# 11. REPORTE DE CLASIFICACIÓN
# ============================================================

print(
    classification_report(
        y_test,
        y_pred,
        target_names=datos.target_names
    )
)

              precision    recall  f1-score   support

   malignant       0.98      0.98      0.98        42
      benign       0.99      0.99      0.99        72

    accuracy                           0.98       114
   macro avg       0.98      0.98      0.98       114
weighted avg       0.98      0.98      0.98       114



In [ ]:
# ============================================================
# 12. PROBABILIDADES
# ============================================================

probabilidades = modelo.predict_proba(X_test_scaled)

print("Primeros 5 casos:")
print()

for i in range(5):
    print(
        f"Caso {i+1}: "
        f"Maligno = {probabilidades[i][0]:.2%} | "
        f"Benigno = {probabilidades[i][1]:.2%}"
    )

Primeros 5 casos:

Caso 1: Maligno = 100.00% | Benigno = 0.00%
Caso 2: Maligno = 0.00% | Benigno = 100.00%
Caso 3: Maligno = 99.36% | Benigno = 0.64%
Caso 4: Maligno = 46.65% | Benigno = 53.35%
Caso 5: Maligno = 100.00% | Benigno = 0.00%


¿ Qué pasó con el desbalanceo?

El desbalanceo existe, pero no es suficientemente extremo como para impedir que la Regresión Logística funcione bien, y además estamos evaluando con varias métricas, no solamente con accuracy.

En este dataset hay aproximadamente:

357 casos benignos
212 casos malignos

Equivale aproximadamente a 63% vs. 37%. Hay desbalance, pero no es un caso extremo como 99% vs. 1%.

Lo importante: ¿qué pasaría si el modelo simplemente eligiera la clase mayoritaria?

En clasificación, accuracy sola puede engañarnos.


En este problema, un falso negativo (FN) significa:

El modelo predijo "benigno" cuando realmente era "maligno".

Ese error puede ser mucho más relevante que un falso positivo.

Por eso conviene mirar:

* Precision
* Recall
* F1-score
* Matriz de confusión

¿Y por qué usamos `stratify=y`?

Esta línea:

train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)

es importante.

`stratify=y` hace que entrenamiento y prueba mantengan aproximadamente la misma proporción de clases que el dataset original.

Sin `stratify`, podríamos tener, por azar, una distribución diferente entre train y test.

Entonces, ¿hay que corregir el desbalanceo?

No necesariamente.

Detectar desbalanceo no significa automáticamente aplicar SMOTE, undersampling o oversampling.

Primero hay que evaluar qué tan desbalanceadas están las clases y qué impacto tiene sobre el modelo.